### DISCUSSION TITLE
Won't be able to create Object definition for the new Master Category.

### DISCUSSION THREAD

**Discussion by komalm** [Sunbird-Knowlg/Community#198] [General Discussions]
Hello @AmiableAnil @krgauraw @pallakartheekreddy

**Issue**: Won't be able to create Object definition for the new Master Category.

**Summary**: We onboarded Event APIs on our server and we are able to use CRUD APIs. We want to make the event a part of the course. When an event is created the default objectType is Event and hence we want to create an object category definition with targetObjectType as the event

We have already created a master category using the following curl

```
curl --location --globoff '{{host}}/api/object/category/v1/create' \
--header 'Content-Type: application/json' \
--header 'Authorization: {{auth}}' \
--data '{
    "request": {
        "objectCategory": {
            "name": "Event",
            "description": "Event"
        }
    }
}'
```

`Response is : obj-cat:event`

Now using this obj-cat we are trying to create the object category definition using the following curl

```
curl --location --globoff '{{host}}/api/object/category/definition/v1/create' \
--header 'Content-Type: application/json' \
--header 'Authorization: {{auth}}' \
--data '{
    "request": {
        "objectCategoryDefinition": {
            "categoryId": "obj-cat:event",
            "targetObjectType": "Event",
            "objectMetadata": {
                "config": {},
                "schema": {
                    "properties": {
                        "trackable": {
                            "type": "object",
                            "properties": {
                                "enabled": {
                                    "type": "string",
                                    "enum": [
                                        "Yes",
                                        "No"
                                    ],
                                    "default": "Yes"
                                },
                                "autoBatch": {
                                    "type": "string",
                                    "enum": [
                                        "Yes",
                                        "No"
                                    ],
                                    "default": "No"
                                }
                            },
                            "default": {
                                "enabled": "Yes",
                                "autoBatch": "No"
                            },
                            "additionalProperties": false
                        },
                        "monitorable": {
                            "type": "array",
                            "items": {
                                "type": "string",
                                "enum": [
                                    "progress-report",
                                    "score-report"
                                ]
                            }
                        },
                        "credentials": {
                            "type": "object",
                            "properties": {
                                "enabled": {
                                    "type": "string",
                                    "enum": [
                                        "Yes",
                                        "No"
                                    ],
                                    "default": "Yes"
                                }
                            },
                            "default": {
                                "enabled": "Yes"
                            },
                            "additionalProperties": false
                        },
                        "userConsent": {
                            "type": "string",
                            "enum": [
                                "Yes",
                                "No"
                            ],
                            "default": "No"
                        }
                    }
                }
            },
            "languageCode": []
        }
    }
}'
```

Response Is - 

```
{
    "id": "api.object.category.definition.create",
    "ver": "3.0",
    "ts": "2024-04-15T06:00:30ZZ",
    "params": {
        "resmsgid": "f6745555-d949-474e-b364-629306277667",
        "msgid": null,
        "err": "CLIENT_ERROR",
        "status": "failed",
        "errmsg": "Validation Errors"
    },
    "responseCode": "CLIENT_ERROR",
    "result": {
        "messages": [
            "Metadata targetObjectType should be one of: [Content, Collection, Asset, ItemSet, Question, QuestionSet]"
        ]
    }
}
```

Can you please assist here in solving this issue? Like adding the Metadata targetObjectType as an Event in the above list.

cc @vaivk369

**Comment by pallakartheekreddy** [comment]
@komalm You need to update `schema.json` of objectcategorydefinition. Because currently `targetObjectType` added above values as enum you need to update that enum values with `Event`.  Please check below `schema.json` for your reference
https://github.com/Sunbird-Knowlg/knowledge-platform/blob/master/schemas/objectcategorydefinition/1.0/schema.json#L28C1-L29C1

**Reply by komalm** [reply]
Hello @pallakartheekreddy,

I wanted to provide an update regarding the issue we've been facing. Despite following the instructions provided and updating the code as suggested, we are still encountering the same issue.

To recap, here are the steps we took:

1. Cloned the content service into our repository.
2. Added the Event value to the schema.
3. Updated the Git URL in the configuration of the Jenkins job for the content service.
4. Built and deployed the content service again.


Unfortunately, the issue persists despite these efforts.

cc @vaivk369, @Venkatesh-Dayani23

**Reply by pallakartheekreddy** [reply]
@komalm Can you share the schema changes what you did? And did you upload the schema after your changes? Can you please provide error details as well.

**Reply by komalm** [reply]
Hello @pallakartheekreddy,

Here is the change [link](https://github.com/Venkatesh-Dayani23/knowledge-platform/blob/release-5.2.0_test/schemas/objectcategorydefinition/1.0/schema.json#L37) 

We built and deployed the same after making the necessary modifications, as I mentioned in the steps.

Having the same issue even after the updates were deployed (Are you requesting this error? or would you prefer another?)

```
{
    "id": "api.object.category.definition.create",
    "ver": "3.0",
    "ts": "2024-04-16T06:39:56ZZ",
    "params": {
        "resmsgid": "e6decc45-8955-4d57-a36e-82daa3854558",
        "msgid": null,
        "err": "CLIENT_ERROR",
        "status": "failed",
        "errmsg": "Validation Errors"
    },
    "responseCode": "CLIENT_ERROR",
    "result": {
        "messages": [
            "Metadata targetObjectType should be one of: [Content, Collection, Asset, ItemSet, Question, QuestionSet]"
        ]
    }
}
```
cc @vaivk369, @Venkatesh-Dayani23

**Reply by pallakartheekreddy** [reply]
@komalm I got it you built and deployed `content-service`. But we made changes in schema, right? We need to upload these modified schemas to blob/S3 (blob if we are using Azure/ S3 bucket if we use AWS).

**Reply by komalm** [reply]
@pallakartheekreddy 

our private DevOps repo: [release-5.1.0/ansible/inventory/dev/KnowledgePlatform/common.yml](https://app.slack.com/client/TNG794XDZ/release-5.1.0/ansible/inventory/dev/KnowledgePlatform/common.yml)

We find the following path:
kp_schema_base_path: "{{proto}}://{{sunbird_public_storage_account_name}}.[blob.core.windows.net/{{plugin_container_name}}/schemas/local](http://blob.core.windows.net/%7B%7Bplugin_container_name%7D%7D/schemas/local)"
we uploaded the new schema file here. and also restarted the content service. If we open the file in the browser we are able to see the changes.

But after executing the create object def curl we are still facing the following issue/error

```
{
    "id": "api.object.category.definition.create",
    "ver": "3.0",
    "ts": "2024-04-16T10:16:10ZZ",
    "params": {
        "resmsgid": "7287ec65-fea7-4627-9f4f-fafef941203a",
        "msgid": null,
        "err": "CLIENT_ERROR",
        "status": "failed",
        "errmsg": "Validation Errors"
    },
    "responseCode": "CLIENT_ERROR",
    "result": {
        "messages": [
            "Metadata targetObjectType should be one of: [Content, Collection, Asset, ItemSet, Question, QuestionSet]"
        ]
    }
}
```

**Reply by pallakartheekreddy** [reply]
@komalm Can we get into a call to check this issue?

**Reply by komalm** [reply]
@pallakartheekreddy Can we schedule a meeting for 5:00 today?

**Reply by pallakartheekreddy** [reply]
@komalm Okay. Can you please send meeting invite?

**Reply by komalm** [reply]
Thank you @pallakartheekreddy,

After restarting the telemetry service, it is working.

**Reply by pallakartheekreddy** [reply]
@komalm Small correction it is Taxonomy service.
